solver/llbsolver/provenance/types/types.go

```diff
@@ -102,15 +178,81 @@ type Environment struct {
 	Platform string `json:"platform"`
 }
 
-type ProvenanceMetadata struct {
-	slsa02.ProvenanceMetadata
-	BuildKitMetadata BuildKitMetadata `json:"https://mobyproject.org/buildkit@v1#metadata,omitempty"`
-	Hermetic         bool             `json:"https://mobyproject.org/buildkit@v1#hermetic,omitempty"`
-}
-
 type BuildKitMetadata struct {
 	VCS      map[string]string                  `json:"vcs,omitempty"`
 	Source   *Source                            `json:"source,omitempty"`
 	Layers   map[string][][]ocispecs.Descriptor `json:"layers,omitempty"`
 	SysUsage []*resourcestypes.SysSample        `json:"sysUsage,omitempty"`
 }
+
+type BuildKitComplete struct {
+	Parameters  bool `json:"parameters"`
+	Environment bool `json:"environment"`
+	Materials   bool `json:"materials"`
+}
+
+func ConvertSLSA02ToSLSA1(p02 *ProvenancePredicateSLSA02) *ProvenancePredicateSLSA1 {
+	if p02 == nil {
+		return nil
+	}
+
+	var resolvedDeps []slsa1.ResourceDescriptor
+	for _, m := range p02.Materials {
+		resolvedDeps = append(resolvedDeps, slsa1.ResourceDescriptor{
+			URI:    m.URI,
+			Digest: m.Digest,
+		})
+	}
+
+	buildDef := ProvenanceBuildDefinitionSLSA1{
+		ProvenanceBuildDefinition: slsa1.ProvenanceBuildDefinition{
+			BuildType:            p02.BuildType,
+			ResolvedDependencies: resolvedDeps,
+		},
+		ExternalParameters: ProvenanceExternalParametersSLSA1{
+			ConfigSource: ProvenanceConfigSourceSLSA1{
+				URI:    p02.Invocation.ConfigSource.URI,
+				Digest: p02.Invocation.ConfigSource.Digest,
+				Path:   p02.Invocation.ConfigSource.EntryPoint,
+			},
+			Parameters:  p02.Invocation.Parameters,
+			BuildConfig: p02.BuildConfig,
+		},
+		InternalParameters: ProvenanceInternalParametersSLSA1{
+			Environment: p02.Invocation.Environment,
+		},
+	}
+
+	var meta *ProvenanceMetadataSLSA1
+	if p02.Metadata != nil {
+		meta = &ProvenanceMetadataSLSA1{
+			BuildMetadata: slsa1.BuildMetadata{
+				InvocationID: p02.Metadata.BuildInvocationID,
+				StartedOn:    p02.Metadata.BuildStartedOn,
+				FinishedOn:   p02.Metadata.BuildFinishedOn,
+			},
+			BuildKitMetadata: p02.Metadata.BuildKitMetadata,
+			Hermetic:         p02.Metadata.Hermetic,
+			Completeness:     BuildKitComplete(p02.Metadata.Completeness),
+			Reproducible:     p02.Metadata.Reproducible,
+		}
+	}
+
+	runDetails := ProvenanceRunDetailsSLSA1{
+		ProvenanceRunDetails: slsa1.ProvenanceRunDetails{
+			Builder: slsa1.Builder{
+				ID: p02.Builder.ID,
+				// TODO: handle builder components versions
+				// Version: map[string]string{
+				// 	"buildkit": version.Version,
+				// },
```

Comment:
Didn't set the builder components versions. I can't find any real use cases in the wild atm. Maybe setting BuildKit version would be fine but as BuildKit has backward compatibility guarantee, it should not affect reproducibility.